Modules/IO/ImageBase/include/itkImageFileWriter.h

```diff
@@ -240,58 +241,26 @@ class ITK_TEMPLATE_EXPORT ImageFileWriter : public ProcessObject
 /** Convenience function for writing an image. */
 template <typename TImage>
 ITK_TEMPLATE_EXPORT void
-WriteImage(const TImage * image, const char * filename, bool compress = false)
+WriteImage(TImage image, const std::string & filename, bool compress = false)
 {
-  using WriterType = ImageFileWriter<TImage>;
-  typename WriterType::Pointer writer = WriterType::New();
+  static_assert(std::is_pointer<TImage>::value || mpl::IsSmartPointer<TImage>::Value,
+                "WriteImage requires a raw pointer or SmartPointer.");
+
+  using ImagePointerType = typename std::conditional<std::is_pointer<TImage>::value,
+                                                     SmartPointer<const typename std::remove_pointer<TImage>::type>,
+                                                     const TImage>::type;
+
+  using ImageType = typename std::remove_const<typename ImagePointerType::ObjectType>::type;
```

Comment:
You might have retrieved the ImageType directly from the image parameter by `decltype(*image)`. Just a suggestion! You'd probably still need remove_const, and also `std::remove_reference`!